Simple Custom Operator API, V0

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #98440

This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

zou3519 committed:
Simple Custom Operator API, V0

This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal right now is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing).

Things that will come in the near-term (before 2.1):
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal right now is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing).

Things that will come in the near-term (before 2.1):
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal right now is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing).

Things that will come in the near-term (before 2.1):
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing).

Things that will come in the near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing), which will come in a future PR
since this one is getting long.

Things that will come in the near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around an OpOverload that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the only blocker is adding
Entrypoint 3 (CustomOp.from_existing), which will come in a future PR
since this one is getting long.

Things that will come in the near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

tugsbayasgalan commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+        # Bypass torch.ops.* and directly do OperatorHandle::callBoxed.
+        # Using torch.ops.* is a bit of a pain (it can be slow and it has lifetime
+        # issues from caching operators that make testing CustomOp difficult).
+        result = _C._dispatch_call_boxed(self._ophandle, *args, **kwargs)
```
Would this forbid us to access the overloaded op directly?

zou3519 replied:
Yes, the user is not allowed to access self._ophandle directly. Note that the user can still do torch.ops.custom.foo (or whatever) if they wanted, but I don't think there should be a reason for the user to do this

tugsbayasgalan commented on test/test_python_dispatch.py:
```diff
@@ -352,6 +363,179 @@ def sqsum(a: SymInt, b: SymInt):
+            return x.sin()
+
+        x = torch.randn(3)
+        result = foo(x)
```
Just wondering, can we directly invoke the overload? Previously we could do like:
```
torch.ops.custom_lib.foo
```

zou3519 replied:
Invoking torch.ops.custom_lib.foo is still supported, though I'm not sure why one would want to grab it directly and invoke it. Do you see use cases for this?

tugsbayasgalan commented on test/test_proxy_tensor.py:
```diff
@@ -1596,23 +1597,23 @@ def f(args, kwargs, extra_args, extra_kwargs):
 
 class TestProxyTensorOpInfo(TestCase):
-    @ops(op_db, allowed_dtypes=(torch.float,))
+    @ops(op_db + custom_op_db, allowed_dtypes=(torch.float,))
```
Can you paste the generated graph from make_fx here? Just wanted to see how the end graph looks like.

zou3519 replied:
```
import torch
from torch.testing._internal.custom_op_db import numpy_mul
from functorch import make_fx

def f(x, y):
    return numpy_mul(x, y)


x = torch.randn(3)
y = torch.randn(2, 1)

gm = make_fx(f, tracing_mode='symbolic')(x, y)
print(gm.code)
```
produces:
```
def forward(self, x_1, y_1):
    numpy_mul = torch.ops._torch_testing.numpy_mul.default(x_1, y_1);  x_1 = y_1 = None
    return numpy_mul
```

zou3519 commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+"""
+
+
+SUPPORTED_DEVICE_TYPES = ("cpu", "cuda")
```
`meta` felt special enough to me that I gave it its own method, no real reason other than that

bdhirsh replied:
Looking forward to the day when a user with a `PrivateUse1` backend wants to update this list 😛 (no action required)

ezyang commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -852,7 +852,12 @@ class OpInfo:
-        self._original_opinfo_args = asdict(self).copy()
+        try:
+            self._original_opinfo_args = asdict(self).copy()
+        except TypeError:
```
Is there a way to do this without catching ALL TypeErrors? Easy way to accidentally suppress too much

zou3519 replied:
I can rewrite the CustomOp OpInfos to use lambdas, e.g. `lambda *args, **kwargs: CustomOp(*args, **kwargs)`. Lambdas can be deep-copied but the CustomOp object cannot

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -0,0 +1,112 @@
+numpy_cube = CustomOp.define('_torch_testing::numpy_cube(Tensor x) -> (Tensor, Tensor)')
+
+@numpy_cube.impl('cpu')
+@numpy_cube.impl('cuda')
```
Huh, weren't you going to support `@numpy_cube.impl('cpu', 'cuda')` syntax?

Minor nit: these examples are a little awkward because, actually, the implementation here would work for ALL devices, not just cpu/cuda.

zou3519 replied:
Can do, but it would probably look more like `@numpy_cube.impl('cpu')` or `@numpy_cube.impl(['cpu', 'cuda'])` because I want to leave space for more positional arguments in the future

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -0,0 +1,112 @@
+
+@numpy_mul.impl_meta()
+def numpy_mul_meta(x, y):
+    return x * y
```
This meta is not actually right 😛

ezyang replied:
Yeah, you're going to get the stock multiply meta behavior which is too flexible

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -0,0 +1,112 @@
+def to_numpy(tensor):
+    return tensor.cpu().numpy()
+
+numpy_cube = CustomOp.define('_torch_testing::numpy_cube(Tensor x) -> (Tensor, Tensor)')
```
OK, imma bikeshed this more.

The downside to implementing the object this way is `numpy_cube` will not have the correct `__name__` and `__module__` stored on it, because obviously CustomOp.define has no idea where its result is going to be bound. The way we handled this in primtorch is to define a stub function and then decorate it, e.g.,

```
@_make_elementwise_unary_reference(ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT)
def abs(a):
    return prims.abs(a)
```

instead of

```
abs = _make_elementwise_unary_reference(ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT, prims.abs)
```

This is troublesome for your use case, because your custom ops don't really have a mandatory function body. Or do they? Forcing a meta implementation doesn't seem like the worst thing in the world.

ezyang replied:
An extra benefit to doing it this way, is the function you feed into the decorator can provide the namespace (assuming we have some suitable mangling scheme from Python module names to TS namespaces, which I think we should!) and the function name, so you don't have to repeat yourself here.

```
@CustomOp.define_meta('(Tensor x) -> (Tensor, Tensor)')
def numpy_cube(x):
    return x.clone(), x.clone()
```

And if you want to get really fancy, we could derive the schema from mypy type annotations:

```
@CustomOp.define_meta()
def numpy_cube(x: Tensor) -> Tuple[Tensor, Tensor]:
    return x.clone(), x.clone()
```

zou3519 replied:
Let me think about this some more. Some initial responses:

> because your custom ops don't really have a mandatory function body. Or do they? Forcing a meta implementation doesn't seem like the worst thing in the world.

If we go this route, I'm leaning towards the user defining a dummy function instead of us having a default:
```py
@CustomOp.define('(Tensor x) -> (Tensor, Tensor)')
def numpy_cube(x):
    pass
```
or take the route of a default. Forcing everyone to have a meta implementation seems good for PT2, but has the downside of that it's not always possible to write the meta implementation (data-dependent rules need to go through a FakeTensor registration, unless we combine the FakeTensor and meta registrations).

> (assuming we have some suitable mangling scheme from Python module names to TS namespaces, which I think we should!) and the function name, so you don't have to repeat yourself here.

I've thought about the mangling the python module name to TS namespace. The downside is that the user (if they're working with the FX graphs or need to use the raw torch.library API) needs to know about the mangling scheme. But maybe this isn't too bad if we document what is going on

ezyang replied:
OK, let's do placeholder. I suggest the idiomatic style should be ellipses rather than pass.

> I've thought about the mangling the python module name to TS namespace. The downside is that the user (if they're working with the FX graphs or need to use the raw torch.library API) needs to know about the mangling scheme. But maybe this isn't too bad if we document what is going on. The pros are that the user really doesn't need to worry about namespace collision and we take care of namespacing for them

Yeah, let's define a permanent mangling scheme (it probably can be pretty simple; just need a prefix like py, an encoding scheme for dot, and an escape for literal occurrence of dot.)

zou3519 replied:
> Yeah, let's define a permanent mangling scheme (it probably can be pretty simple; just need a prefix like py, an encoding scheme for dot, and an escape for literal occurrence of dot.)

@ezyang our options seem like either we do the mangling, or we teach schema about nested namespaces. I'm not sure how difficult the latter is right now, but there are some other things we need to do if we go the mangling approach:
- I want to put some hook into torch_dispatch so that when torch.ops.PYmylibraryDOTcustom_ops_V0.foo.default gets returned to Python, we actually get back mylibrary.custom_ops.foo instead of the OpOverload and also have the demangled form show up in the FX graph (so that it's easier to manipulate). Are we OK with this? 
- People who are trying to override the operation from C++ or using torch.library need to work with the mangled name ("PYmylibraryDOTcustom_ops_V0::foo"), unless we want to let them somehow use the unmangled name.

Also, if we are inferring the name of the operator from the python function, then, we need to figure out some initial plan to deal overloads (at least the out= variant) for now, since that will be the first thing that people ask us for.

ezyang replied:
Latter puts you into getting TorchScript on board, which just seems like a lot of pain.

> I want to put some hook into torch_dispatch so that when torch.ops.PYmylibraryDOTcustom_ops_V0.foo.default gets returned to Python, we actually get back mylibrary.custom_ops.foo instead of the OpOverload and also have the demangled form show up in the FX graph (so that it's easier to manipulate). Are we OK with this?

Yes, this sounds good. Since you're eschewing OpOverload anyway you probably wanted to do this anyway, so that you get your new object type. But whatever API you expose needs to be compatible with OpOverload...

> People who are trying to override the operation from C++ or using torch.library need to work with the mangled name ("PYmylibraryDOTcustom_ops_V0::foo"), unless we want to let them somehow use the unmangled name.

This seems fine

> Also, if we are inferring the name of the operator from the python function, then, we need to figure out some initial plan to deal overloads (at least the out= variant) for now, since that will be the first thing that people ask us for.

Seems fine, we can have a convention for out= variant overload name.

ezyang commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+torch.library/TORCH_LIBRARY, which require deep understanding of PyTorch internals,
+as well as being more comprehensive than autograd.Function, which only supports
+implementing gradient computation and vmap rules.
+"""
```
Define an `__all__` for this module.

ezyang commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+# We will not let users register CustomOps with anything that could look like
+# PyTorch internals to avoid confusion.
+RESERVED_NS = {
+    "prims",
```
You need `prim` too. But I also think we should not let users pick namespaces and instead base them off of Python module hierarchy.

bdhirsh replied:
Inferring based off of python module hierarchy sounds pretty good (and fits the theme of "CustomOp" is a layer of abstraction above `torch.Library` with fewer toggles/knobs)

ezyang commented:
Mechanically, I have no complaints. It's all API bikeshedding lol.

zou3519 commented:
> Mechanically, I have no complaints. It's all API bikeshedding lol.

API Bikeshedding is great, please keep it coming

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+
+CustomOp is positioned as being safer and easier to use than
+torch.library/TORCH_LIBRARY, which require deep understanding of PyTorch internals,
+as well as being more comprehensive than autograd.Function, which only supports
```
is it worth mentioning that CustomOp will (probably) be more amenable to tracing / using with torch.compile compared to autograd.Function?

zou3519 replied:
Yup!

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+
+    # Coming in the future. Requires us to have correct logic for
+    # the ADInplaceOrView key
+    if schema.kind() != SchemaKind.functional:
```
Do you intend to make `CustomOp` (semi)-public before the functionalization support? Maybe we should just add a big warning somewhere in the docs so people know that performing input mutations in your custom op can cause problems.

bdhirsh commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,292 @@
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        # TODO(rzou): RAII guard should be contextmanager, or else bad things will happen
+        guard = _C._AutoDispatchBelowAutograd()
```
probably a dumb q: what bad things can happen if we just use a try/finally block here?

zou3519 replied:
The object would not get deleted if we accidentally created a reference cycle with it, but if we're careful when writing the code then we should be fine. So this is more of an idiomatic thing where python encourages people to use context managers instead of rely on RAII, especially because other implementations of python (not cpython) do not need to guarantee this.

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -352,6 +363,179 @@ def sqsum(a: SymInt, b: SymInt):
+        foo = CustomOp.define('custom::foo(Tensor x) -> Tensor')
+        x = torch.randn(3, requires_grad=True)
+        with self.assertRaisesRegex(RuntimeError, 'Autograd has not been implemented'):
+            foo(x)
```
The C++ "not implemented" autograd kernel is a bit lazier - you can run the forward without error, but install a "NotImplemented" node in the autograd graph - so it will only yell at you if you later try to backprop through your op. Do you think we should do the same for these custom ops?

zou3519 replied:
Yes, we should install the C++ "not implemented" autograd kernel. My plan was to handle this in a follow-up because this PR is getting long by itself and it's not trivial to use the C++ "not implemented" autograd kernel (because there are two of them, one for Autograd and one for ADInplaceAndView) and I need to figure out how to call them from Python

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -0,0 +1,126 @@
+def numpy_cube_meta(x):
+    return x.clone(), x.clone()
+
+@CustomOp.define('(Tensor x, Tensor y) -> Tensor', ns='_torch_testing')
```
minor API nit: `@custom_op(schema)` seems more direct, and the user doesn't really care about the CustomOp class.

ezyang replied:
Or maybe, you should create a namespace, and then use that to define operators on the namespace.

zou3519 replied:
I agree that `@custom_op(schema)` is more direct, but in the context of the three entrypoints (only the first exists right now), it makes sense that they are accessed similarly:
- Entrypoint 1: decorator CustomOp.define (for your new custom ops)
- Entrypoint 2: decorator CustomOp.from_decomp (for your custom operators that are composed of only PyTorch operations, we can autogenerate rules for them)
- Entrypoint 3: function CustomOp.from_existing (pass in an operator registered with TORCH_LIBRARY/torch.library here)

Maybe we can combine entrypoint 1 and 2 into a `@custom_op(schema, *, ns, decomp=False)` and actually use the function the user provided to us as the decomp when implementations don't exist and come up with a succinct name for Entrypoint 3? ("as_custom_op"?)

ezyang replied:
yeah I would just add kwargs to `custom_op`, or just have another function entirely

For preexisting function, you actually probably want to hang it on torch.ops, TBH. Because you don't actually want the custom_op outer function.

ezyang commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,333 @@
+
+def validate_namespace(ns: str) -> None:
+    if ns not in RESERVED_NS:
+        return
```
Check that the namespace doesn't have periods in it?

ezyang commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,333 @@
+        raise NotImplementedError(
+            f"NYI: CustomOp.define does not support function schema with no Tensor inputs. Got: {schema}"
+        )
+    # Just seems weird so banning for now
```
What about if it's side-effectful? That seems legit.

ezyang replied:
oh, you checked this above

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 commented:
Just a quick FYI for everyone else: we decided to not go with auto-generating the namespace (for now, at least). Here's the chain of thought:
- let's say we wanted to mangle the entire namespace (e.g. my_library.bar.baz) into one that can be understood by the C++ dispatcher.
- Then, ideally we would hide the mangled namespace -- the main use case for this is FX graph transforms or torch_dispatch, and we don't want users to use torch.ops.PYmy_librarybarbazM10_3_3.foo
- This is OK if we return our own object from torch_dispatch, not torch.ops.PYmy_librarybarbazM10_3_3.foo
- But the implication is that our own object needs to be compatible with the OpOverload/OpOverloadPacket, and existing systems (ProxyTensor, FakeTensor) actually rely on these having names that look like "bar::foo".

I think we can autogenerate the namespace if we get nested namespaces (which is kind of difficult) or somehow special case namespaces from Python (e.g. flag them in the Dispatcher with "from_python" or something). But to keep things simple for now we ask the user to manually pass the namespace, and tell them best practices (e.g. the namespace they choose should probably just be their top-level module name).

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 commented on torch/_custom_op.py:
```diff
@@ -0,0 +1,346 @@
+        >>> # We need to provide the decorator a "prototype function"
+        >>> # (a function with Python ellipses as the body).
+        >>> @custom_op('(Tensor x) -> Tensor')
+        >>> def numpy_sin(x):
```
Unfortunately, no :P. Prototype function is just there so we can get the name/module

I can add some check (that the number of arguments and the name of the arguments are the same?) if you think it would be nicer

soulitzer replied:
Oh wait this is the API you proposed in your doc initially lol, why did we switch to this new version?

zou3519 replied:
@ezyang, some feedback that we should go back to the original API lol.

@soulitzer, the reason why we switched to the new version is the `__module__`/`__name__` aren't set "correctly". That is, if you do the following, then `numpy_sin.__name__` is None and `numpy_sin.__module__` is torch._custom_op.

```py
numpy_sin = custom_op('numpy_sin(Tensor x) -> Tensor')
```

Now, the problem is mostly that the user may expect `numpy_sin` to behave like a function (it will be interposable by torch_function), but it doesn't really look like a function (it doesn't have `__name__` or `__module__` set like a function). 

At this point it's a tradeoff discussion. We can confuse the user one way or the other, so what is less confusing?

zou3519 replied:
A secondary reason why we switched, but I don't think will actually happen in the future, is that we may be able to infer a schema from the prototype function's signature. But I wouldn't count on this, because there's not an existing way to capture alias information in the schema, and if we ever need to express more information in the schema we may have a problem:
```
@custom_op(ns='_torch_testing')
def numpy_sin(x: Tensor) -> Tensor
    ...
```

soulitzer replied:
Ah thanks for the context.

Re: inferring the schema, since view or inplace is not common, could we make specifying the schema optional unless we detect that the user tries to do either of those things? That would make it possible to infer the schema for most cases right?

soulitzer replied:
> IMO it is better if we drill into the user that it is possible for their schema to be wrong (and have them read through a doc page on how to write a schema and what it consists of), instead of us detecting it ourselves, because it's a bit difficult to detect ourselves:

Yeah I agree with this

Given that the secondary reason is unlikely to materialize. Would it resolve the issue if we just set __name__ and __module__ ourselves?

zou3519 replied:
> Would it resolve the issue if we just set __name__ and __module__ ourselves?

Well, we'd need to ask our users to set __name__ and __module__, right? Inside of the `custom_op` function, when we construct the CustomOp object, we do not know the __module__ of `custom_op`'s caller

zou3519 replied:
@soulitzer I am only going to check that the number of arguments and the name of the arguments are correct. Before we make this API public, we will decide if we want to autogenerate the schema from the function signature or not. In the latter case, we may need additional checks: that the type annotations in the function signature are compatible with the schema types.

lezcano replied:
I don't know if we want to go there, but note that we could set the `__name__` by looking at the schema, and we could set the module by looking at `inspect.getmodule(inspect.stack()[1][0])`. Now, it's reasonable if you guys think that this is a bit too much.

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 committed:
Update on "Simple Custom Operator API, V0"


This PR introduces CustomOp, a wrapper around a dispatcher operator that allows
users to define custom operators. It adds the skeleton for CustomOp and
some very simple behavior: as of this PR:
- one can create a CustomOp for an operator that does not have inplace or aliasing
- give it CPU/CUDA and Meta implementations
- and trace it into a graph via make_fx.

The design follows
https://docs.google.com/document/d/19Uc5OUCA187q9BZggJb70RT2ZoSTDoG5QQkJkZwd25M/edit
Concretely, we implement the following things mentioned in the doc in this PR:
- Entrypoint 1 (CustomOp.define, creating a new custom operator)
- impl (to define device-specific code) and impl_meta (to define meta
formulas)

The goal for the short term is to get the code to a state where it can be trialed
by the export folks. On top of this PR, the blockers are:
- adding Entrypoint 3 (CustomOp.from_existing)
- adding a way to do data-dependent shape formulas
These will come in future PRs since this one is getting long.

Things that will come in the longer-near-term (before 2.1):
- adding the other entrypoints mentioned in the doc (2 & 3)
- more safety checks and better error messages
- support for views and mutation
- support for defining autograd formulas
- support for functionalization
- making this API public (it's private right now).

Test Plan:
- added a new test case, TestCustomOp. It mostly tests a bunch of error
cases.
- added OpInfos for custom operators and hooked these up to
test_proxy_tensor to test that they work with make_fx. These custom
operators were based off of the ones in the autograd_function_db.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
@zou3519 sorry, I have to revert your PR. It leads to internal failures. Here you can find more details - [D45066168](https://www.internalfb.com/diff/D45066168).
